### PR TITLE
#22 [DataMatrix] Support for rectangular DataMatrix

### PR DESCRIPTION
**Pull request by Int32Overflow**
#22

### CODE REVIEW

**Comment by Int32Overflow** [comment]
Do you have time to take a look at it? Unfortunately I still have some problems and I can't find the errors. 
There must be something wrong with the algorithm. I have already found one or the other mistake and fixed it.

grafik

**Comment by Int32Overflow** [comment]
@huysentruitw 
grafik

**Comment by BenMakesGames** [comment]
has this PR been abandoned? I'm also interested in this feature, and would be willing to look into fixing the unit tests.

File: tests/Barcoder.Tests/DataMatrix/DataMatrixEncoderTests.cs [hunk 1]
```diff
@@ -180,6 +180,154 @@ public void Encode_Gs1ModeEnabled_ValidGs1Content_NonPreDefinedApplicationIdenti
...
+            TestEncode_FixedSize(12, 26, content, expectedDataBits);
+        }
+
+        [Fact]
+        public void Encode_12x36()
+        {
+            var content = "Abc12345Hij!klm";
+            var expectedDataBits = ImageStringToBools(@"
+                #.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.
+                #.##....#.##.##.#######......###....
```
Comment (BenMakesGames) [inline tests/Barcoder.Tests/DataMatrix/DataMatrixEncoderTests.cs]: ^ this test data is incorrect on this line; the very last character should be a `#`, and not a `.`.

making that changes fixes this unit test.

**Comment by Int32Overflow** [comment]
> has this PR been abandoned? I'm also interested in this feature, and would be willing to look into fixing the unit tests.

I am also still interested and would be happy if we could do it together.

**Comment by BenMakesGames** [comment]
I'm looking at the 16x48 test now. When overlaying the DataMatrix as generated by this library vs this online tool: http://generator.onbarcode.com/online-data-matrix-barcode-generator.aspx I am getting different results.

The gray areas show where there are differences in the images; there are many:

image

Both codes scan as "AbcDef1234567Abc1" using a DataMatrix app I got for my phone, but I assume that's because of the error correction features.

I don't actually know much about how these codes are generated, so I don't have any ideas as to what's going on. I'm going to start reading up on DataMatrix, and see if I can figure anything out.

**Comment by Int32Overflow** [comment]
@BenMakesGames Yes Test Encode_16x36 is green now. Juppiii

But 16x36 and 16x48 are failing 👎

**Comment by Int32Overflow** [comment]
@BenMakesGames I'm using this library at the moment:
https://www.bcgen.com/demo/IDAutomationStreamingDataMatrix.aspx?MODE=0&D=AbcdF12345ZY#123&PFMT=27&PT=F&X=0.5&O=0&LM=0.2

But I think it has a problem with the '#' character.

Maybe we should create one without '#' character

**Comment by BenMakesGames** [comment]
`#` is a special character in URLs; it just needs to be encoded as `%23`:

https://www.bcgen.com/demo/IDAutomationStreamingDataMatrix.aspx?MODE=0&D=AbcdF12345ZY%23123&PFMT=27&PT=F&X=0.5&O=0&LM=0.2

**Comment by BenMakesGames** [comment]
I've been comparing the code in this library to that of other libraries, and nothing obvious is standing out to me.

One option could be to remove/comment out 16x36 and 16x48 from the list of supported codes; let someone down the line solve those :P

I happen to only need 8x32 for the project I'm working on. I'm going to discuss with my coworkers how much time I should devote to this. It'd be nice if I could spend the time to give back to the open source community, but there's other tickets I need to work on.

It's nearing the end of the work day for me, so I'm going to stop working on it for now. I _might_ look into this more over the weekend, but I make no promises...

**Comment by Int32Overflow** [comment]
> I've been comparing the code in this library to that of other libraries, and nothing obvious is standing out to me.

I also did not see anything

**Comment by huysentruitw** [comment]
Thanks to both of you, I was able to crack the code. There was an error in the padding algorithm ! 🎉

Many thanks for this addition 🙌🙌

**Comment by huysentruitw** [comment]
@Int32Overflow Please update the `AddPadding` method to:

```C#
        internal static byte[] AddPadding(byte[] data, int toCount)
        {
            var result = new List<byte>(data);
            if (result.Count < toCount)
                result.Add(129);

            while (result.Count < toCount)
            {
                int r = ((149 * (result.Count + 1)) % 253) + 1;
                int tmp = 129 + r;
                if (tmp > 254)
                    tmp -= 254;

                result.Add((byte)tmp);
            }

            return result.ToArray();
        }
```

**Comment by BenMakesGames** [comment]
oh, wow! I had tried rewriting the padding algorithm, since the codes seemed to have more errors in the bottom-right. I guess my rewrite must have also been bad :P

**Comment by Int32Overflow** [comment]
> @Int32Overflow Please update the `AddPadding` method to:
> 
> ```cs
>         internal static byte[] AddPadding(byte[] data, int toCount)
>         {
>             var result = new List<byte>(data);
>             if (result.Count < toCount)
>                 result.Add(129);
> 
>             while (result.Count < toCount)
>             {
>                 int r = ((149 * (result.Count + 1)) % 253) + 1;
>                 int tmp = 129 + r;
>                 if (tmp > 254)
>                     tmp -= 254;
> 
>                 result.Add((byte)tmp);
>             }
> 
>             return result.ToArray();
>         }
> ```

I will take a look at it.

**Comment by Int32Overflow** [comment]
@Int32Overflow Thank you

**Comment by Int32Overflow** [comment]
@huysentruitw Thank you

**Pull request merged**